pymeasure/adapters/adapter.py

```diff
@@ -34,44 +37,112 @@ class Adapter:
     This class should only be inherited from.
 
     :param preprocess_reply: optional callable used to preprocess strings
-        received from the instrument. The callable returns the processed string.
+        received from the instrument. The callable returns the processed
+        string. This argument is deprecated.
     :param kwargs: all other keyword arguments are ignored.
     """
 
     def __init__(self, preprocess_reply=None, **kwargs):
         self.preprocess_reply = preprocess_reply
         self.connection = None
+        self.log = logging.Logger("Adapter")
+        self.log.addHandler(logging.NullHandler())
+        if preprocess_reply is not None:
+            warn(
+                "Do not use `Adapter.values`, but `Instrument.values` instead. "
+                "Preprocess_reply is a values parameter.",
+                FutureWarning)
 
     def __del__(self):
         """Close connection upon garbage collection of the device"""
-        if self.connection is not None:
+        self.close()
+
+    def close(self):
+        """Close the connection."""
+        try:
             self.connection.close()
+        except AttributeError:
+            pass
 
-    def write(self, command):
-        """ Writes a command to the instrument
+    # Directly called methods. DO NOT OVERRIDE IN SUBCLASS!
+    def write(self, command, **kwargs):
+        """
+        Write a string command to the instrument.
 
-        :param command: SCPI command string to be sent to the instrument
+        The write_termination string is appended, if defined.
+
+        :param command: Command string to be sent to the instrument.
+        :param kwargs: Keyword arguments for the connection itself.
+        """
+        self.log.debug(("WRITE:", command))
+        self._write(command, **kwargs)
+
+    def write_bytes(self, content, **kwargs):
+        """
+        Write the bytes `content` to the instrument.
+
+        :param content: The bytes to write to the instrument.
+        :param kwargs: Keyword arguments for the connection itself.
+        """
+        self.log.debug(("WRITE:", content))
+        self._write_bytes(content, **kwargs)
```

Comment:
Should you log _after_ writing? Otherwise e.g. the PrologixAdapter address command appears in the log in a different order to what goes over the connection?

Reply:
was intended for `write`, but point still stands

Reply:
Right now, the generator executes a task and reads afterwards the accumulated communication. Therefore it does not matter whether write or log(write) is first. It is only important, that the log entries are in the same order as the read/writes

Reply:
> It is only important, that the log entries are in the same order as the read/writes

Indeed, but is that the case?
The flow I reconstructed in my head was:
In `Instrument.write`, we get
* a log message "WRITE: bla"
* a call to Adapter.write, which ultimately calls PrologixAdapter._write, which 
* writes an address command
* logs an address command, and then
* writes "bla".

So, the order is reversed, no?

Reply:
I answered in the other comment.
But yes you are right, and I changed the order